text/0005-is-operator.md

```diff
@@ -0,0 +1,176 @@
+- `User::"alice" is Namespace::User` is false
+
+## Changes required to validator
+
+`e is et` will be typed as `True` if the validator can guarantee that expression `e` is an entity of type `et`, or `False` if the validator can guarantee that `e` is a non-entity type or an entity not of type `et`.
+Otherwise, the validator gives `e is et` type `Bool`.
+
+To improve typing precision, we can extend the current notion of "effects" to track information about entity types.
+Currently, the validator uses effects to know whether an expression is guaranteed to have an attribute.
+In particular, the validator knows that if `e has a` evaluates to true, then expression `e` is guaranteed to have attribute `a`.
+We can do something similar for `is`: when the expression `e is et` evaluates to true, expression `e` is guaranteed to be an entity of type `et`.
+
+## Other changes
+
+The Dafny formalization of the Cedar language, corresponding proofs, and differential testing framework in [cedar-spec](https://github.com/cedar-policy/cedar-spec) will need to be updated with the new `is` operator.
+
+# Drawbacks
+
+1. This is a substantial new feature that will require significant development effort.
+
+2. This change will break existing applications that depend on the structure of policies. For example, if we decide to allow `is` in the policy scope (as shown in the examples above), tools that perform basic "policy slicing" based on the scope will need to decide how to handle the new syntax.
+
+3. The `is` operator makes individual policies easier to understand because the type information is in the policy itself. However, it may make it more difficult to get a global view of how different actions and entity types are used. The schema currently fills this role: it provides a centralized place to provide all typing constraints.
+
+4. The `is` operator may encourage coding practices that go against our recommended "best practices" (more details below).
+
+## Best practice: Using "typed" actions
+
+From the discussion on <https://github.com/cedar-policy/cedar/issues/94>:
+
+> Guidance we have given other customers is that actions should be "typed," in the sense that they include in their name the type of the resource they apply to. So instead of writing
+>
+>     permit(
+>       principal == User::"admin",
+>       action == Action::"view",
+>       resource is Photo
+>     );
+>
+> you would write
+>
+>     permit(
+>       principal == User::"admin",
+>       action == Action::"viewPhoto",
+>       resource
+>     );
+>
+> In your schema, you indicate that the "viewPhoto" action should always be accompanied by a `Photo` for the resource, so the policy doesn't need to say that explicitly (and doesn't need to perform a runtime check).
+
+If we support an `is` operator, then customers may be encouraged to use a single "view" action for multiple resource types, rather than a new "viewX" actions for each resource type X.
+
```

Comment:
Our own motivating example violates this tenet. I think we should change the example so that we satisfy the tenet, but still show that adding `is` makes sense.

Reply:
One think I was wondering while writing up this RFC: why do we have this tenet? What is "bad" about using a generic action type with multiple principal/resource types?

Reply:
Answering the question about what's "bad" about generic action type...  This is a lesson learned through experience. There are two key points:

- The input context can vary when acting on different resource types. For a real-life example from AWS, see the [documentation for S3](https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazons3.html). Make your screen really big so you can see the column in the table named "Condition keys". That's the equivalent of what Cedar calls "context". Notice how the context values can be different when operating on different resource types. If generic actions were used, then the context would be the union of all possible inputs for all resource types, most of them labeled as optional, and it becomes challenging to provide useful validation or documentation for the action.
- It's easy to accidentally overgrant permissions. For example, consider an action like "read" that applies to 10 different resource types. It's relatively easy to slip up and write a policy that accidentally overgrants to too many types of resources. By looking at the policy alone, it's not entirely clear what it gives access to. Even if the `is` operator existed, somebody still needs to remember to use it. In contrast, resource-specific action names are constrained by default, and self-documenting.
- And, if desired, you can still get the same effect of generic action names by decorating each action with attributes like `type=[read|write|list|view|delete]`, and writing policies like `permit(...) when {action.type == "read"}`

Looking at the major cloud providers as a point of reference, AWS uses resource-specific action names. Azure and GCP appear to as well, although structured a bit differently. For example, rather than an RPC-style action name like `getClustersInContainer`, it will be `container.clusters.get`. But, similar effect.